Packs/AWS-SNS-Listener/Integrations/AWSSNSListener/AWSSNSListener.py

```diff
@@ -0,0 +1,255 @@
+
+
+def handle_teams_proxy_and_ssl():
+    proxies = None
+    use_ssl = not PARAMS.get('insecure', False)
+    if not is_demisto_version_ge('8.0.0'):
+        return proxies, use_ssl
+    CRTX_HTTP_PROXY = os.environ.get('CRTX_HTTP_PROXY', None)
+    demisto.debug(f"CRTX_HTTP_PROXY: {CRTX_HTTP_PROXY}")
+    if CRTX_HTTP_PROXY:
+        proxies = {
+            "http": CRTX_HTTP_PROXY,
+            "https": CRTX_HTTP_PROXY
+        }
+        use_ssl = True
+    return proxies, use_ssl
+
+
+PROXIES, USE_SSL = handle_teams_proxy_and_ssl()
+
+
+def valid_sns_message(sns_payload):
+    """
+    Validates an incoming Amazon Simple Notification Service (SNS) message.
+
+    Args:
+        sns_payload (dict): The SNS payload containing relevant fields.
+
+    Returns:
+        bool: True if the message is valid, False otherwise.
+    """
+    # taken from https://github.com/boto/boto3/issues/2508
+    # Can only be one of these types.
+    if sns_payload["Type"] not in ["SubscriptionConfirmation", "Notification", "UnsubscribeConfirmation"]:
+        demisto.error('Not a valid SNS message')
+        return False
+
+    # Amazon SNS currently supports signature version 1 or 2.
+    if sns_payload["SignatureVersion"] not in ["1", "2"]:
+        demisto.error('Not using the supported AWS-SNS SignatureVersion 1 or 2')
+        return False
+
+    # Fields for a standard notification.
+    fields = ["Message", "MessageId", "Subject", "Timestamp", "TopicArn", "Type"]
+
+    # Determine the required fields based on message type
+    if sns_payload["Type"] in ["SubscriptionConfirmation", "UnsubscribeConfirmation"]:
+        fields = ["Message", "MessageId", "SubscribeURL", "Timestamp", "Token", "TopicArn", "Type"]
+
+    # Build the string to be signed.
+    string_to_sign = ""
+    for field in fields:
+        string_to_sign += field + "\n" + sns_payload[field] + "\n"
+
+    # Verify the signature
+    decoded_signature = base64.b64decode(sns_payload["Signature"])
+    try:
+        resp = requests.get(sns_payload["SigningCertURL"], verify=USE_SSL, proxies=PROXIES)
+        resp.raise_for_status()
+        certificate = X509.load_cert_string(resp.text)
+    except Exception as e:
+        demisto.error(f'Exception validating sign cert url: {e}')
+        return False
+
+    public_key = certificate.get_pubkey()
+    # Verify the signature based on SignatureVersion
+    if sns_payload["SignatureVersion"] == "1":
+        public_key.reset_context(md="sha1")
+    else:  # version2
+        public_key.reset_context(md="sha256")
+
+    public_key.verify_init()
+    public_key.verify_update(string_to_sign.encode())
+    verification_result = public_key.verify_final(decoded_signature)
+
+    if verification_result != 1:
+        demisto.error('Signature verification failed.')
+        return False
+
+    demisto.debug('Signature verification succeeded.')
+    return True
+
+
+@app.post('/')
+async def handle_post(request: Request,
```

Comment:
Since the function is very long, it'll be better to add a docstring and/or split the function to a bit more sub-functions to make the code more readable.